### PR TITLE
[DOCS] Docs-only improvements

### PR DESCRIPTION
**Pull request by jaceklaskowski**
…adoc

## What changes were proposed in this pull request?

Use recommended values for row boundaries in Window's scaladoc, i.e. `Window.unboundedPreceding`, `Window.unboundedFollowing`, and `Window.currentRow` (that were introduced in 2.1.0).

## How was this patch tested?

Local build

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 1]
```diff
@@ -22,7 +22,7 @@ import org.apache.spark.sql.Column
 import org.apache.spark.sql.catalyst.expressions._
 
 /**
- * Utility functions for defining window in DataFrames.
+ * Utility functions for defining window in Datasets.
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: These are used with DataFrames, right? At least that's what I have used Window for, am I missing something?

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 2]
```diff
@@ -113,12 +113,12 @@ object Window {
    * Creates a [[WindowSpec]] with the frame boundaries defined,
    * from `start` (inclusive) to `end` (inclusive).
    *
-   * Both `start` and `end` are relative positions from the current row. For example, "0" means
+   * Both `start` and `end` are relative positions to the current row. For example, "0" means
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: I think the right phrasing is: "are positions relative to the current row". The current text is OK IMHO

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 3]
```diff
@@ -113,12 +113,12 @@ object Window {
...
    * "current row", while "-1" means the row before the current row, and "5" means the fifth row
    * after the current row.
    *
-   * We recommend users use `Window.unboundedPreceding`, `Window.unboundedFollowing`,
-   * and `Window.currentRow` to specify special boundary values, rather than using integral
+   * We recommend users to use [[Window.unboundedPreceding]], [[Window.unboundedFollowing]],
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: "We recommend that users use" is correct, but 'that' can be omitted and it's still correct.
I think the backticks are on purpose as many scaladoc refs like this also cause doc failures. At least you need to verify this before changing.

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 4]
```diff
@@ -131,9 +131,9 @@ object Window {
    *   import org.apache.spark.sql.expressions.Window
    *   val df = Seq((1, "a"), (1, "a"), (2, "a"), (1, "b"), (2, "b"), (3, "b"))
    *     .toDF("id", "category")
-   *   df.withColumn("sum",
-   *       sum('id) over Window.partitionBy('category).orderBy('id).rowsBetween(0,1))
-   *     .show()
+   *   val byCategoryOrderedById =
+   *     Window.partitionBy('category).orderBy('id).rowsBetween(Window.currentRow, 1)
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: Why this change? 0 should mean current row.

**Review by srowen — COMMENTED** [review]
I'm not sure any of these changes are necessary?

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala (hunk 1 above)
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: Sure. I'm going to revert the changes. There's a little value in them. I'd rather see the changes approved in general than fight for DataFrame vs Dataset.

p.s. There's no `DataFrame` in Spark SQL which is just a type alias of `Dataset[Row]` -- see https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/package.scala#L46.

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala (hunk 2 above)
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: I'll fix it to be more accurate (that's the purpose of this particular change so the more accurate the merrier). Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala (hunk 4 above)
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: See the above change where Spark devs "recommend that users use" the values by their aliases not their numeric values.

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala (hunk 3 above)
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: Leaving 'that' aside is incorrect -- see http://dictionary.cambridge.org/dictionary/english/recommend where *to* is even highlighted to make the point.

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala (hunk 1 above)
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: It's an alias, yes, but it certainly exists as a user-facing type.

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 5]
```diff
@@ -162,11 +162,11 @@ object Window {
    * Creates a [[WindowSpec]] with the frame boundaries defined,
    * from `start` (inclusive) to `end` (inclusive).
    *
-   * Both `start` and `end` are relative from the current row. For example, "0" means "current row",
+   * Both `start` and `end` are relative to the current row. For example, "0" means "current row",
    * while "-1" means one off before the current row, and "5" means the five off after the
    * current row.
    *
-   * We recommend users use `Window.unboundedPreceding`, `Window.unboundedFollowing`,
+   * We recommend that users use `Window.unboundedPreceding`, `Window.unboundedFollowing`,
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: Either way is correct, it wasn't wrong

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala [hunk 6]
```diff
@@ -200,9 +200,9 @@ object Window {
    * }}}
    *
    * @param start boundary start, inclusive. The frame is unbounded if this is
-   *              the minimum long value (`Window.unboundedPreceding`).
+   *              the minimum long value, i.e. `Window.unboundedPreceding`.
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: Likewise this is effectively identical. I wouldn't make changes like this

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala (hunk 4 above)
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/Window.scala]: Got it, and I think there are also doc examples like this in `Column.scala` and `WindowSpec.scala` that could be similarly improved

**Comment by jaceklaskowski** [comment]
Hey @srowen Would appreciate your looking at the changes again and comments (or merge). Thanks!

**Comment by srowen** [comment]
How about the other files I mentioned? I think they can take similar changes. I think you can roll your other PR into this. They're both kinda misc doc improvements.

**Comment by jaceklaskowski** [comment]
I'm going to merge the two PRs with your comments applied (i.e. excluding changes that are not necessarily doc-only). Thanks a lot for your time, Sean. Appreciate a lot.

**Comment by jaceklaskowski** [comment]
Executed `cd docs && SKIP_PYTHONDOC=1 SKIP_RDOC=1 jekyll serve`  to check the changes and they've seemed fine. I had to fix some extra javadoc-related places to please jekyll.

@srowen Ready to review the changes once more? Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 7]
```diff
@@ -75,7 +75,6 @@ case class WindowSpecDefinition(
       frameSpecification.isInstanceOf[SpecifiedWindowFrame]
 
   override def nullable: Boolean = true
-  override def foldable: Boolean = false
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: I get that this is redundant, or happens to be right now, but I don't think I'd remove it in a docs-only change

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala [hunk 8]
```diff
@@ -26,7 +26,8 @@ import org.apache.spark.sql.types._
 import org.apache.spark.unsafe.types.CalendarInterval
 
 /**
- * Test basic expression parsing. If a type of expression is supported it should be tested here.
+ * Test basic expression parsing.
+ * If the type of an expression is supported it should be tested here.
```
Comment (srowen) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: This is a no-op change, I'd avoid this.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala [hunk 9]
```diff
@@ -60,7 +60,7 @@ import org.apache.spark.util.Utils
  * The builder can also be used to create a new session:
  *
  * {{{
- *   SparkSession.builder()
+ *   SparkSession.builder
```
Comment (srowen) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: Is this for consistency? it also seems not worth changing otherwise

**Review by srowen — CHANGES_REQUESTED** [review]
Most of this looks fine to me, but I have a few questions still about the changes.

File: core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java [hunk 10]
```diff
@@ -52,16 +52,15 @@
  * This class implements sort-based shuffle's hash-style shuffle fallback path. This write path
  * writes incoming records to separate files, one file per reduce partition, then concatenates these
  * per-partition files to form a single output file, regions of which are served to reducers.
- * Records are not buffered in memory. This is essentially identical to
- * {@link org.apache.spark.shuffle.hash.HashShuffleWriter}, except that it writes output in a format
+ * Records are not buffered in memory. It writes output in a format
```
Comment (jaceklaskowski) [inline core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java]: `HashShuffleWriter` is long gone.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 7 above)
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: Correct. Reverting...

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala (hunk 8 above)
Comment (jaceklaskowski) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ExpressionParserSuite.scala]: Almost. I replaced `a` with `the` and added `an` before `expression`.

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala (hunk 9 above)
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: Consistency (and one of the recommended coding styles of mine).

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Looks good, just needs a rebase now

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**